### PR TITLE
feat: new ci

### PR DESCRIPTION
**Pull request by smrz2001**
This PR implements a new CI pipeline using [Dagger](https://dagger.io/).

This allows the various Ceramic component pipelines to be written entirely agnostically of the pipeline service (e.g. GitHub Workflows).

Dagger is still a little rough around the edges so some of the syntax is non-intuitive. However, as they improve their DX, we'll be able to clean this code up as well.

I've decided to keep the actual pipeline [Dagger plan](https://docs.dagger.io/1202/plan) in a [separate repository](https://github.com/3box/pipelinetools) that can be used from other Ceramic repos instead of scattering the code across the repos.

I have intentionally not deprecated the old pipeline (CircleCI, building an image, posting to Lambda, etc.) until the new one can run correctly end-to-end in parallel.

Tasks accomplished with this PR:

- [x] Builds the code
- [x] Runs unit tests (`go` and `js`)
- [x] Builds the Docker image
- [x] Start the Docker image and runs a `curl` request to the `healthcheck` endpoint.
- [x] Pushes the image to DockerHub and ECR
- [x] Queues an event to the job queue

### CODE REVIEW

File: .github/workflows/ci.yml
Comment (gvelez17) [inline .github/workflows/ci.yml:72]: tiny nit but maybe this should read
# Pull the latest pipelinetools. TODO: once the code is stable, use a tagged version

**Comment by gvelez17** [comment]
LGTM if it works!  Two tiny nits/qs above, my main question before merging is what will it do when we merge this and how do we test it?  If we merge this now does it have an immediate effect on the pipelines, do we need to have any kind of eyes on when we merge this to prod?

**Comment by smrz2001** [comment]
> LGTM if it works! Two tiny nits/qs above, my main question before merging is what will it do when we merge this and how do we test it? If we merge this now does it have an immediate effect on the pipelines, do we need to have any kind of eyes on when we merge this to prod?

Until I implement the CD portion, which will be responsible for running deployments, nothing changes for existing pipelines if/when this is merged. PR and dev/rc/main pushes will start running the new CI code and pushing images, that's it.

Eventually, I'd like to enable the new CI as a check that _must_ pass before PRs can be merged but that can come later, once the rest of the flow is fleshed out.

@gvelez17

**Comment by ukstv** [comment]
Do we really need _all_ of the stuff in `universe.dagger.io`?

**Comment by oed** [comment]
> Please ignore the cue.mod folder. This is currently the recommended way to pin dependencies.

The only way to pin dependencies is to include them in your github repo 🤔

**Comment by smrz2001** [comment]
> > Please ignore the cue.mod folder. This is currently the recommended way to pin dependencies.
> 
> The only way to pin dependencies is to include them in your github repo 🤔

Haha, @oed, yeah, it is for now, though the Dagger team is working on explicit dependency management.

I am still not 100% convinced with Dagger but seeing how actively (and responsively) it's being developed, plus the soul-hurting pain I feel thinking about implementing all this directly in Github or CodeBuild is keeping me going in their direction 😅

We could go with a different vebdor-agnostic solution like Jenkins but that would also mean understanding a whole new language (Groovy) and figuring out how to deploy our own Jenkins pipeline, etc.

**Review by v-stickykeys — COMMENTED** [review]
After looking at the other repo I'm not fully seeing the benefit of dagger here if we still run everything with gh runner, and need logic in gh actions for every step we want to do in dagger. Any failure or other conditional logic still has to go in gh actions. This seems like a higher barrier to entry than where we are now (people have to learn both gh actions and dagger) and duplication of code to some extent.

If the paradigm is gh action is just a skeleton for the steps that invoke the actual execution we should allow devs to write each step in any language (bash, python, js, etc.) and invoke that script here.

(Side note: Can we remove these example files or is it useful to have them? cue.mod/pkg/universe.dagger.io/examples/todoapp/public/logo512.png)

File: .github/workflows/ci.yml
Comment (smrz2001) [reply .github/workflows/ci.yml:19]: No, these are `dagger.#Secret` type variables and masked through the execution.

Comment (smrz2001) [reply .github/workflows/ci.yml:19]: Github will keep them masked in its runtime throughout the workflow execution.

**Comment by smrz2001** [comment]
@ukstv @oed @v-stickykeys, based on Val's comments, I've cleaned up the code further.

Only took me a few hours, but now, _everything_ is covered by the Dagger plan. All GitHub (or another service) needs to do is install Dagger and run `dagger do build`.

All committed dependencies have been removed.

**Review by v-stickykeys — APPROVED** [review]
👍

**Pull request merged**
